--- a/types/scene/surface.c	2024-04-14 03:06:16.194056481 -0300
+++ b/types/scene/surface.c	2024-04-14 03:03:50.956829757 -0300
@@ -6,6 +6,7 @@
 #include <wlr/types/wlr_linux_drm_syncobj_v1.h>
 #include <wlr/types/wlr_presentation_time.h>
 #include <wlr/util/transform.h>
+#include <wlr/render/timeline.h>
 #include "types/wlr_scene.h"
 
 static void handle_scene_buffer_outputs_update(
@@ -138,9 +139,11 @@
 		pixman_region32_intersect_rect(&opaque, &opaque, 0, 0, width, height);
 	}
 
+	
+	struct wlr_render_timeline *timeline = calloc(1, sizeof(*timeline));
 	if (width <= 0 || height <= 0) {
 		wlr_scene_buffer_set_buffer(scene_buffer, NULL);
-		wlr_scene_buffer_set_wait_timeline(scene_buffer, NULL, 0);
+		wlr_scene_buffer_set_wait_timeline(scene_buffer, timeline, 0);
 		pixman_region32_fini(&opaque);
 		return;
 	}
@@ -163,7 +166,7 @@
 
 	pixman_region32_fini(&opaque);
 
-	struct wlr_render_timeline *wait_timeline = NULL;
+	struct wlr_render_timeline *wait_timeline = calloc(1, sizeof(*wait_timeline));
 	uint64_t wait_point = 0;
 	struct wlr_linux_drm_syncobj_surface_v1_state *syncobj_surface =
 		wlr_linux_drm_syncobj_v1_get_surface_state(surface);
