--- a/backend/drm/libliftoff.c
+++ b/backend/drm/libliftoff.c
@@ -407,7 +407,7 @@
 		struct wlr_drm_connector *conn = state->connectors[i].connector;
 		struct wlr_drm_crtc *crtc = conn->crtc;
 
-		int ret = liftoff_output_apply(crtc->liftoff, req, flags);
+		int ret = liftoff_output_apply(crtc->liftoff, req, flags, NULL);
 		if (ret != 0) {
 			wlr_drm_conn_log(conn, test_only ? WLR_DEBUG : WLR_ERROR,
 				"liftoff_output_apply failed: %s", strerror(-ret));
